sdks/java/harness/src/test/java/org/apache/beam/fn/harness/FnApiDoFnRunnerTest.java

```diff
@@ -505,10 +516,226 @@ public void testSideInputIsAccessibleForDownstreamCallers() throws Exception {
     assertEquals(stateData, fakeClient.getData());
   }
 
+  private static class TestTimerfulDoFn extends DoFn<KV<String, String>, String> {
+    private static final TupleTag<String> mainOutput = new TupleTag<>("mainOutput");
+    private static final TupleTag<String> additionalOutput = new TupleTag<>("output");
+
+    @TimerId("event")
+    private final TimerSpec eventTimerSpec = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    @TimerId("processing")
+    private final TimerSpec processingTimerSpec = TimerSpecs.timer(TimeDomain.PROCESSING_TIME);
+
+    @ProcessElement
+    public void processElement(
+        ProcessContext context,
+        @TimerId("event") Timer eventTimeTimer,
+        @TimerId("processing") Timer processingTimeTimer) {
+      context.output("main" + context.element().getKey());
+      eventTimeTimer.set(context.timestamp().plus(1L));
+      processingTimeTimer.offset(Duration.millis(2L));
+      processingTimeTimer.setRelative();
+    }
+
+    @OnTimer("event")
+    public void eventTimer(
+        OnTimerContext context,
+        @TimerId("event") Timer eventTimeTimer,
+        @TimerId("processing") Timer processingTimeTimer) {
+      context.output("event");
+      eventTimeTimer.set(context.timestamp().plus(11L));
+      processingTimeTimer.offset(Duration.millis(12L));
+      processingTimeTimer.setRelative();
+    }
+
+    @OnTimer("processing")
+    public void processingTimer(
+        OnTimerContext context,
+        @TimerId("event") Timer eventTimeTimer,
+        @TimerId("processing") Timer processingTimeTimer) {
+      context.output("processing");
+      eventTimeTimer.set(context.timestamp().plus(21L));
+      processingTimeTimer.offset(Duration.millis(22L));
+      processingTimeTimer.setRelative();
+    }
+  }
+
+  @Test
+  public void testTimers() throws Exception {
+    dateTimeProvider.setDateTimeFixed(10000L);
+
+    Pipeline p = Pipeline.create();
+    PCollection<KV<String, String>> valuePCollection =
+        p.apply(Create.of(KV.of("unused", "unused")));
+    PCollection<String> outputPCollection =
+        valuePCollection.apply(TEST_PTRANSFORM_ID, ParDo.of(new TestTimerfulDoFn()));
+
+    SdkComponents sdkComponents = SdkComponents.create();
+    sdkComponents.registerEnvironment(Environment.getDefaultInstance());
+    // Note that the pipeline translation for timers creates a loop between the ParDo with
+    // the timer and the PCollection for that timer. This loop is unrolled by runners
+    // during execution which we redo here manually.
+    RunnerApi.Pipeline pProto = PipelineTranslation.toProto(p, sdkComponents);
+    String inputPCollectionId = sdkComponents.registerPCollection(valuePCollection);
+    String outputPCollectionId = sdkComponents.registerPCollection(outputPCollection);
+    String eventTimerInputPCollectionId = "pTransformId/ParMultiDo(TestTimerful).event";
```

Comment:
Do we need to ensure that this is consistent with how proto translation/runner operations work? If so, how do we keep it in sync?

Reply:
Proto translation should rarely change so I don't expect this to be much of an issue but in reality once we have end to end validates runner tests we should really delete a lot of the translation/execution tests since validates runner tests are much simpler to write and maintain.